### PR TITLE
feat(hotkey): implement the RPC handler in hotkey_collector

### PR DESCRIPTION
**Pull request by Smityz**
implement the RPC handler in hotkey_collector

### CODE REVIEW

File: src/server/hotkey_collector.h [hunk 1]
```diff
@@ -19,6 +19,8 @@
 
 #include <dsn/utility/string_view.h>
 #include <dsn/dist/replication/replication_types.h>
+#include <dsn/dist/replication/replica_base.h>
+#include <dsn/dist/fmt_logging.h>
```
Comment (acelyc111) [inline src/server/hotkey_collector.h:23]: Move this line to cpp file

File: src/server/hotkey_collector.cpp [hunk 2]
```diff
@@ -16,14 +16,34 @@
...
+hotkey_collector::hotkey_collector(dsn::replication::hotkey_type::type hotkey_type,
+                                   dsn::replication::replica_base *r_base)
+    : replica_base(r_base), _state(hotkey_collector_state::STOPPED), _hotkey_type(hotkey_type)
+{
+}
+
 void hotkey_collector::handle_rpc(const dsn::replication::detect_hotkey_request &req,
                                   dsn::replication::detect_hotkey_response &resp)
 {
+    if (req.action == dsn::replication::detect_action::START) {
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:33]: Use `switch..case..`

File: src/server/hotkey_collector.cpp [hunk 3]
```diff
@@ -16,14 +16,34 @@
...
+        std::string err_hint;
+        if (start_detect(err_hint)) {
+            resp.err = dsn::ERR_OK;
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:36]: You can move this line at first in this function, then it's not needed to reassign it again and again.

File: src/server/hotkey_collector.cpp [hunk 4]
```diff
@@ -33,5 +53,40 @@ void hotkey_collector::capture_raw_key(const dsn::blob &raw_key, int64_t weight)
 
 void hotkey_collector::capture_hash_key(const dsn::blob &hash_key, int64_t weight) {}
 
+bool hotkey_collector::start_detect(std::string &err_hint)
+{
+    auto now_state = _state.load();
+    switch (_state.load()) {
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:59]:
```suggestion
    switch (now_state) {
```

File: src/server/hotkey_collector.cpp [hunk 5]
```diff
@@ -33,5 +53,40 @@ void hotkey_collector::capture_raw_key(const dsn::blob &raw_key, int64_t weight)
...
+    case hotkey_collector_state::COARSE_DETECTING:
+    case hotkey_collector_state::FINE_DETECTING:
+        err_hint = fmt::format("still detecting {} hotkey, state is {}",
+                               dsn::enum_to_string(_hotkey_type),
+                               enum_to_string(now_state));
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:64]: Unify to add or remove `dsn::`

File: src/server/hotkey_collector.cpp [hunk 6]
```diff
@@ -33,5 +53,40 @@ void hotkey_collector::capture_raw_key(const dsn::blob &raw_key, int64_t weight)
...
+        ddebug_replica(err_hint);
+        return false;
+    case hotkey_collector_state::STOPPED:
+        // TODO: (Tangyanzhao) start coarse detecting
+        _state.store(hotkey_collector_state::COARSE_DETECTING);
+        ddebug_replica("starting to detect {} hotkey", dsn::enum_to_string(_hotkey_type));
+        return true;
+    default:
+        err_hint = "invalid collector state";
+        ddebug_replica(err_hint);
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:80]: It's not possible to reach here, I think you can assert here.

File: src/server/hotkey_collector.cpp (hunk 5 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:64]: change the namespace may lead to compile failed
`dsn::replication::hotkey_type::type` and `pegasus::hotkey_collector_state` are defined in different namespaces

File: src/server/hotkey_collector.h [hunk 7]
```diff
@@ -76,8 +77,15 @@ class hotkey_collector
                     /*out*/ dsn::replication::detect_hotkey_response &resp);
 
 private:
-    std::unique_ptr<internal_collector_base> _internal_collector;
+    bool is_detecting_now();
```
Comment (levy5307) [inline src/server/hotkey_collector.h:80]: inline

File: src/server/hotkey_collector.cpp [hunk 8]
```diff
@@ -17,21 +17,53 @@
...
 {
 }
 
-// TODO: (Tangyanzhao) implement these functions
 void hotkey_collector::handle_rpc(const dsn::replication::detect_hotkey_request &req,
                                   dsn::replication::detect_hotkey_response &resp)
 {
+    switch (req.action) {
+    case dsn::replication::detect_action::START:
+        if (is_detecting_now()) {
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:42]: you can encapsulate into a new function for each case.

File: src/server/hotkey_collector.cpp [hunk 9]
```diff
@@ -17,21 +17,53 @@
...
+            return_err_resp(resp,
+                            fmt::format("still detecting {} hotkey, state is {}",
+                                        dsn::enum_to_string(_hotkey_type),
+                                        enum_to_string(_state.load())));
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:46]: add return;

File: src/server/hotkey_collector.cpp [hunk 10]
```diff
@@ -49,5 +81,36 @@ void hotkey_collector::capture_hash_key(const dsn::blob &hash_key, int64_t weigh
 
 void hotkey_collector::analyse_data() { _internal_collector->analyse_data(); }
 
+bool hotkey_collector::is_detecting_now()
+{
+    return (_state.load() == hotkey_collector_state::COARSE_DETECTING ||
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:86]:
```
   auto currnet_state = _state.load();
   return (current_state == hotkey_collector_state::COARSE_DETECTING ||
            current_state == hotkey_collector_state::FINE_DETECTING);
```

File: src/server/hotkey_collector.cpp (hunk 8 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:42]: According to this comment https://github.com/apache/incubator-pegasus/pull/621#discussion_r506004399, I combined these functions

File: src/server/hotkey_collector.cpp [hunk 11]
```diff
@@ -17,21 +17,37 @@
...
 {
+    switch (req.action) {
+    case dsn::replication::detect_action::START:
+        on_start_detect(resp);
+        return;
+    case dsn::replication::detect_action::STOP:
+        on_stop_detect(resp);
+        return;
+    default:
+        resp.err = dsn::ERR_INVALID_VERSION;
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:48]: Add some log

File: src/server/hotkey_collector.cpp [hunk 12]
```diff
@@ -49,5 +65,50 @@ void hotkey_collector::capture_hash_key(const dsn::blob &hash_key, int64_t weigh
...
+        _state.store(hotkey_collector_state::COARSE_DETECTING);
+        resp.err = dsn::ERR_OK;
+        hint = fmt::format("starting to detect {} hotkey", dsn::enum_to_string(_hotkey_type));
+        ddebug_replica(hint.c_str());
+        return;
+    default:
+        hint = "invalid collector state";
+        resp.err = dsn::ERR_INVALID_VERSION;
+        resp.__set_err_hint(hint.c_str());
+        ddebug_replica(hint.c_str());
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:99]:
```suggestion
        dwarn_replica(hint.c_str());
```

File: src/server/hotkey_collector.cpp [hunk 13]
```diff
@@ -49,5 +65,50 @@ void hotkey_collector::capture_hash_key(const dsn::blob &hash_key, int64_t weigh
...
+    auto now_state = _state.load();
+    std::string hint;
+    switch (now_state) {
+    case hotkey_collector_state::COARSE_DETECTING:
+    case hotkey_collector_state::FINE_DETECTING:
+        resp.err = dsn::ERR_SERVICE_ALREADY_EXIST;
+        hint = fmt::format("still detecting {} hotkey, state is {}",
+                           dsn::enum_to_string(_hotkey_type),
+                           enum_to_string(now_state));
+        ddebug_replica(hint.c_str());
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:79]:
```suggestion
        ddebug_replica(hint);
```

Comment (acelyc111) [reply src/server/hotkey_collector.cpp:79]: The same to other places.

File: src/server/hotkey_collector.cpp [hunk 14]
```diff
@@ -17,21 +17,38 @@
...
+    case dsn::replication::detect_action::START:
+        on_start_detect(resp);
+        return;
+    case dsn::replication::detect_action::STOP:
+        on_stop_detect(resp);
+        return;
+    default:
+        resp.err = dsn::ERR_INVALID_VERSION;
+        resp.__set_err_hint("can't find this detect_action");
+        dwarn_replica("can't find this detect_action");
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:50]: Add error details, i.e. `req.action`, into hint message and log, then admins know what happend.

File: src/server/hotkey_collector.cpp [hunk 15]
```diff
@@ -49,5 +66,50 @@ void hotkey_collector::capture_hash_key(const dsn::blob &hash_key, int64_t weigh
...
+        _state.store(hotkey_collector_state::COARSE_DETECTING);
+        resp.err = dsn::ERR_OK;
+        hint = fmt::format("starting to detect {} hotkey", dsn::enum_to_string(_hotkey_type));
+        ddebug_replica(hint);
+        return;
+    default:
+        hint = "invalid collector state";
+        resp.err = dsn::ERR_INVALID_VERSION;
+        resp.__set_err_hint(hint);
+        dwarn_replica(hint);
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:100]: Better to assert here, coz the state is internal enum, default case will not happen.

File: src/server/hotkey_collector.cpp (hunk 14 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:50]: if `req.action` is an unknown enum, I can't transform it into a string to log it.

**Review by acelyc111 — APPROVED** [review]

File: src/server/hotkey_collector.h [hunk 16]
```diff
@@ -63,10 +63,11 @@ class internal_collector_base;
 //    |                    |  |                     Hotkey                         |
 //    +--------------------+  +----------------------------------------------------+
 
-class hotkey_collector
+class hotkey_collector : public dsn::replication::replica_base
```
Comment (hycdong) [inline src/server/hotkey_collector.h:66]: Is it necessary to inherit `replica_base` class? Will hotkey_collector use other functions except log-printing functions like `ddebug_replica`?

File: src/server/hotkey_collector.cpp [hunk 17]
```diff
@@ -17,21 +17,39 @@
...
+        on_start_detect(resp);
+        return;
+    case dsn::replication::detect_action::STOP:
+        on_stop_detect(resp);
+        return;
+    default:
+        std::string hint = fmt::format("{}: can't find this detect action", req.action);
+        resp.err = dsn::ERR_INVALID_VERSION;
+        resp.__set_err_hint(hint);
+        dwarn_replica(hint);
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:51]: It is better to use error level log here.

File: src/server/hotkey_collector.cpp [hunk 18]
```diff
@@ -49,5 +67,51 @@ void hotkey_collector::capture_hash_key(const dsn::blob &hash_key, int64_t weigh
...
 void hotkey_collector::analyse_data() { _internal_collector->analyse_data(); }
 
+void hotkey_collector::on_start_detect(dsn::replication::detect_hotkey_response &resp)
+{
+    auto now_state = _state.load();
+    std::string hint;
+    switch (now_state) {
+    case hotkey_collector_state::COARSE_DETECTING:
+    case hotkey_collector_state::FINE_DETECTING:
+        resp.err = dsn::ERR_SERVICE_ALREADY_EXIST;
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:77]: Maybe `ERR_INVALID_STATE` is better, if you think ERR_INVALID_STATE is not suitable to show the error, you can define a new error_code.
The error_code problem is also in other cases in this switch condition.
Besides, I wonder what will rpc client act if it receives this error?

File: src/server/hotkey_collector.cpp [hunk 19]
```diff
@@ -49,5 +67,51 @@ void hotkey_collector::capture_hash_key(const dsn::blob &hash_key, int64_t weigh
...
+        hint = fmt::format("still detecting {} hotkey, state is {}",
+                           dsn::enum_to_string(_hotkey_type),
+                           enum_to_string(now_state));
+        ddebug_replica(hint);
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:81]: I think warning or error level may be better, the following log-printing also has this problem.

File: src/server/hotkey_collector.cpp [hunk 20]
```diff
@@ -49,5 +67,51 @@ void hotkey_collector::capture_hash_key(const dsn::blob &hash_key, int64_t weigh
...
+        resp.err = dsn::ERR_INVALID_VERSION;
+        resp.__set_err_hint(hint);
+        dwarn_replica(hint);
+        dassert(false, "invalid collector state");
+    }
+}
+
+void hotkey_collector::on_stop_detect(dsn::replication::detect_hotkey_response &resp)
+{
+    _state.store(hotkey_collector_state::STOPPED);
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:108]: There are no state check here, it means detection can be stopped in any states.

File: src/server/hotkey_collector.h (hunk 16 above)
Comment (Smityz) [reply src/server/hotkey_collector.h:66]: `hotkey_collector` will not use other functions related to `replica_base` without `ddebug_replica`

File: src/server/hotkey_collector.cpp (hunk 20 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:108]: Yes, it is a mandatory termination function

**Review by hycdong — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]

**Pull request merged**